backends/tfhe-cuda-backend/cuda/src/aes128/aes_utilities.h

```diff
@@ -0,0 +1,432 @@
+        this->carry_lut->get_degree(0), this->carry_lut->get_max_degree(0),
+        params.glwe_dimension, params.polynomial_size, params.message_modulus,
+        params.carry_modulus, carry_lambda, allocate_gpu_memory);
+    auto active_streams_carry_lut = streams.active_gpu_subset(num_blocks);
+    this->carry_lut->broadcast_lut(active_streams_carry_lut);
+  }
+
+  void release(CudaStreams streams) {
+    this->and_lut->release(streams);
+    delete this->and_lut;
+    this->and_lut = nullptr;
+
+    this->flush_lut->release(streams);
+    delete this->flush_lut;
+    this->flush_lut = nullptr;
+
+    this->carry_lut->release(streams);
+    delete this->carry_lut;
+    this->carry_lut = nullptr;
+  }
+};
+
+template <typename Torus> struct int_aes_trivial_buffers {
+  CudaRadixCiphertextFFI *trivial_1_bit;
+  Torus *d_trivial_scalars_zero;
+  Torus *h_trivial_scalars_zero;
+
+  int_aes_trivial_buffers(CudaStreams streams, const int_radix_params &params,
+                          bool allocate_gpu_memory, uint64_t &size_tracker) {
+
+    constexpr uint32_t NUM_TRIVIAL_BLOCKS = 1;
+
+    Torus h_trivial_one[] = {1};
+    Torus *d_trivial_scalars = (Torus *)cuda_malloc_with_size_tracking_async(
+        sizeof(Torus), streams.stream(0), streams.gpu_index(0), size_tracker,
+        allocate_gpu_memory);
+    if (allocate_gpu_memory) {
+      cuda_memcpy_async_to_gpu(d_trivial_scalars, h_trivial_one, sizeof(Torus),
+                               streams.stream(0), streams.gpu_index(0));
+    }
+
+    this->trivial_1_bit = new CudaRadixCiphertextFFI;
+    create_zero_radix_ciphertext_async<Torus>(
+        streams.stream(0), streams.gpu_index(0), this->trivial_1_bit,
+        NUM_TRIVIAL_BLOCKS, params.big_lwe_dimension, size_tracker,
+        allocate_gpu_memory);
+
+    if (allocate_gpu_memory) {
+      set_trivial_radix_ciphertext_async<Torus>(
+          streams.stream(0), streams.gpu_index(0), this->trivial_1_bit,
+          d_trivial_scalars, h_trivial_one, NUM_TRIVIAL_BLOCKS,
+          params.message_modulus, params.carry_modulus);
+    }
+    if (allocate_gpu_memory) {
+      cuda_drop_async(d_trivial_scalars, streams.stream(0),
+                      streams.gpu_index(0));
+    }
+
+    h_trivial_scalars_zero = (Torus *)calloc(NUM_TRIVIAL_BLOCKS, sizeof(Torus));
+    d_trivial_scalars_zero = (Torus *)cuda_malloc_with_size_tracking_async(
+        sizeof(Torus), streams.stream(0), streams.gpu_index(0), size_tracker,
+        allocate_gpu_memory);
+    if (allocate_gpu_memory) {
+      cuda_memset_async(d_trivial_scalars_zero, 0, sizeof(Torus),
+                        streams.stream(0), streams.gpu_index(0));
+    }
+  }
+
+  void release(CudaStreams streams, bool allocate_gpu_memory) {
+    if (allocate_gpu_memory) {
+      cuda_drop_async(this->d_trivial_scalars_zero, streams.stream(0),
+                      streams.gpu_index(0));
+    }
+    free(this->h_trivial_scalars_zero);
```

Comment:
You need to synchronize the stream before this free. Otherwise you won't know if there is some transfer happening to/from the device.